tests/test_ga4_conversion_window.py

```diff
@@ -0,0 +1,168 @@
+        streams_to_selected_fields = {stream: set() for stream in self.streams_to_test()}
+        self.select_streams_and_fields(conn_id, test_catalogs, streams_to_selected_fields)
+
+        # set state to ensure conversion window is used
+        today_datetime = dt.strftime(dt.utcnow(), self.BOOKMARK_FORMAT)
+        initial_state = {
+            'bookmarks': {
+                stream: { os.getenv('TAP_GA4_PROPERTY_ID'): {'last_report_date': today_datetime}}
+                for stream in self.streams_to_test()
+            }, 'currently_syncing': None
+        }
+
+        menagerie.set_state(conn_id, initial_state)
+
+        # Run a sync
+        sync_job_name = runner.run_sync_mode(self, conn_id)
+
+        # Verify the tap and target do not throw a critical error
+        exit_status = menagerie.get_exit_status(conn_id, sync_job_name)
+        menagerie.verify_sync_exit_status(self, exit_status, sync_job_name)
+
+        # Verify tap replicates through today by check state
+        final_state = menagerie.get_state(conn_id)
+        self.assertDictEqual(final_state, initial_state)
+
+        # Verify converstion window is used
+        synced_messages_by_stream = runner.get_records_from_target_output()
+        for stream in self.streams_to_test():
+            with self.subTest(stream=stream):
+
+                stream_name = self.get_stream_name(stream)
+                expected_replication_key = next(iter(self.expected_metadata().get(stream_name).get(self.REPLICATION_KEYS)))
+                if not expected_replication_key:
+                    LOGGER.warn(f"********** Failed to get replication key for stream: {stream_name}")
+                    continue
+
+                replication_dates =[row.get('data').get(expected_replication_key) for row in
+                                    synced_messages_by_stream.get(stream, {'messages': []}).get('messages', [])
+                                    if row.get('data')]
+
+                if replication_dates:
+                    conversion_window_int = int(self.conversion_window)
+                    oldest_date = self.parse_date(min(replication_dates))
+                    oldest_possible_date = (dt.utcnow().replace(
+                        hour=0, minute=0, second=0, microsecond=0) - timedelta(days=conversion_window_int))
+                    start_date = self.parse_date(self.start_date)
+
+                    # Verify start date is before the oldest replicated record per test set up
+                    self.assertGreater(oldest_date, start_date)
+                    # Verify the oldest record is not older than conversion window
+                    self.assertGreaterEqual(oldest_date, oldest_possible_date)
+                    # Verify minimal gap between oldest record and conversion_window
+                    # allow a 1 day gap for stability / rollover? TODO
+                    self.assertTrue(timedelta(days=1) > (oldest_date - oldest_possible_date))
```

Comment:
This looks good. Are we guaranteed to have data every day of the previous 90 days.  Otherwise this might not be stable.  I think as long as the date is greater than the next smaller conversion window and after the start_date and conversion window this might be good enough.  For instance for a 90 day conversion window more than 60 days ago means we aren't using the 60 day window, for the 60 day window more than 30 means the same thing.  So I think with 30, 60 and 90 we might be able to extend the 1 up to around 29 days and still be pretty safe.

Reply:
For the working stream we appear to have 1 record per day dating back 90 days during test runs.  I didn't verify that definitively but whatever I set the conversion window to that's how many records come back so it's probably a good guess, I'll take a closer look at the records next run to confirm.  If that holds then I think the current asserts are more rigourous and until we add a stream that has data gaps we can keep them.  If we see gaps we can move to the "next smaller conversion window plus 1 day" approach and use this 29 days-ish approach for the min window of 30.

Reply:
Confirmed 1 record per day for content_group_report and today I'm getting 1 less than conversion window records per sync (maybe time of day type issue so the current logic seems good). Given quota constraints, I'm tempted to just remove the custom report from this test as it seems like it may be orthogonal with respect to conversion window let's discuss.